feat: OpenShift Quickstart guide with Kong Mesh (#6572)
Kong/docs.konghq.com: app/_src/mesh/features/openshift-quickstart.md

---
title: Get started with Red Hat OpenShift and Kong Mesh
content_type: tutorial
description: This guide explains how to get started on Kong Mesh with Red Hat OpenShift.
---

This is a quickstart to get you up and running with Kong Mesh in standalone mode on OpenShift.

This tutorial does not require a license, Kong Mesh can start in evaluation mode, and limits you 5 dataplanes/sidecars to use. Just enough dataplanes to get comfortable with the product.

For OpenShfit, we'll be spinning up a ROSA 4.12 cluster. This tutorial does also assume some base level knowledge of OpenShift.

This tutorial will cover:

* How to use the Red Hat Certified Kong Mesh Images
* How to implement the required OpenShift SecurityContextConstraints (SCC) for the kong-mesh sidecar

> * Deploy and join the Kong for Kubernetes Ingress Controller (KIC) to the mesh

Author: Why do we need to deploy KIC in this guide?

Reviewer: We deploy KIC so that the user can reach those microservice from curl or the browser and validate visually everything is in working order.

* Deploy a sample application, bookinfo, on the mesh and validate it's all working

## Prerequisites
* [ROSA CLI installed](https://docs.openshift.com/rosa/rosa_install_access_delete_clusters/rosa_getting_started_iam/rosa-installing-rosa.html) or another OpenShift 4.12 cluster with the ability to create LoadBalancer type Kubernetes Services
* [kubectl CLI installed](https://kubernetes.io/docs/tasks/tools/)
* [OpenShift CLI installed](https://docs.openshift.com/container-platform/4.11/cli_reference/openshift_cli/getting-started-cli.html)
* [Helm 3 installed](https://helm.sh/docs/intro/install/)

## Install ROSA

1. Create System Variables:
    ```bash
    CLUSTER_NAME=kong-mesh-2
    REGION=us-west-2
    ```

1. Create a small ROSA cluster:
    ```bash
    rosa create cluster --cluster-name=$CLUSTER_NAME --region=$REGION --multi-az=false --version 4.12.13
    ```

1. When the Cluster install is complete, create a cluster-admin user:
    ```bash
    rosa create admin --cluster $CLUSTER_NAME
    ```

1. Validate you can login to the cluster via the credentials provided by the rosa cli stdout. Once login is successful you can proceed to the next step.

> ## Install Kong Mesh in Standalone Mode

Author: Why do we install in standalone mode instead of a different mode?

Reviewer: for simplicity, but converting the instructions to a zone, or mesh manager zone is more appropriate

1. Create the `kong-mesh-system` namespace:
    ```bash
    kubectl create namespace kong-mesh-system
    ```

> 1. Create the image pull secret:

Reviewer: the certified images sit in redhat's image repository and requires customers to authenticate in order to pull down the images

    ```bash
    kubectl create secret docker-registry rh-registry-secret -n kong-mesh-system \
        --docker-server=registry.connect.redhat.com \
        --docker-username=<username> \
        --docker-password=<password> \
        --docker-email=<email>
    ```

> 1. Add nonroot-v2 to job service accounts:

Author: What does this do?

Reviewer: This is escalating the permissions of the service accounts (kong-mesh-install-crds, kong-mesh-patch-ns-job, etc) to nonroot-v2

    ```bash
    oc adm policy add-scc-to-user nonroot-v2 system:serviceaccount:kong-mesh-system:kong-mesh-install-crds
    oc adm policy add-scc-to-user nonroot-v2 system:serviceaccount:kong-mesh-system:kong-mesh-patch-ns-job 
    oc adm policy add-scc-to-user nonroot-v2 system:serviceaccount:kong-mesh-system:kong-mesh-pre-delete-job
    ```

1. Grab the latest helm chart:
    ```bash
    helm repo add kong-mesh https://kong.github.io/kong-mesh-charts
    ```

1. Install Kong Mesh:
    ```bash
    helm upgrade -i kong-mesh kong-mesh/kong-mesh --version 2.2.0 -f kong-mesh/values.yaml -n kong-mesh-system
    ```

1. Open a second terminal and port-forward to reach the mesh GUI:
    ```bash
    kubectl port-forward svc/kong-mesh-control-plane -n kong-mesh-system 5681:5681
    ```

    You should be able to reach the Kong-Mesh UI at `http://localhost:5681/gui`.

> 1. Last, do some prep work for the sidecar itself so sidecars will startup successfully. Apply the kong-mesh-sidecar scc and corresponding container patches:

Author: How does this prep them to start successfully?

Reviewer: These are two new custom files we need to bring into the doc or host somewhere. 

1. [kong-mesh-sidecar-scc](https://github.com/Kong/kong-mesh-quickstart-openshift/blob/main/kong-mesh/kong-mesh-sidecar-scc.yaml) : 
2. [container-patch](https://github.com/Kong/kong-mesh-quickstart-openshift/blob/main/kong-mesh/container-patch.yaml)

In Openshift, by default, all pods and containers are given the restricted security context constraint (SCC) but this is insufficient for the mesh sidecars (containers). The sidecars need a slightly escalated permissions, these permissions are defined in the sidecar-scc manifest. 

Step 1 - create a custom SCC, but this manifest is not associated to any pod container or sidecar in Openshift. 

Step 2 - defines how to patch the sidecars with the needed SCC. 

There is a third step. Step 3 - line 133 - all application pods that will have sidecars, need to have their service accounts given permission to use the custom SCC.

    kubectl apply -f kong-mesh/container-patch.yaml 
    ```

## Deploy KIC on Mesh

1. First we're going to deploy KIC as our ingress controller:
    ```bash
    kubectl create namespace kong 
    kubectl label namespace kong kuma.io/sidecar-injection=enabled
    oc adm policy add-scc-to-user kong-mesh-sidecar system:serviceaccount:kong:kong-kong
    ```

1. Grab the latest chart:
    ```bash
    helm repo add kong https://charts.konghq.com
    helm repo update
    ```

1. Then run the install:
    ```bash
    helm install kong kong/kong -n kong \
    --set ingressController.installCRDs=false \
    --set podAnnotations."kuma\.io/mesh"=default \
    --set podAnnotations."kuma\.io/gateway"=enabled
    export PROXY_IP=$(kubectl get -o jsonpath="{.status.loadBalancer.ingress[0].hostname}" service -n kong kong-kong-proxy)
    curl -i $PROXY_IP
    ```

    The expected output is:
    ```json
    HTTP/1.1 404 Not Found
    Date: Tue, 02 May 2023 15:40:05 GMT
    Content-Type: application/json; charset=utf-8
    Connection: keep-alive
    Content-Length: 48
    X-Kong-Response-Latency: 0
    Server: kong/3.2.2

    {"message":"no Route matched with those values"}%
    ```

> ## Deploy Bookinfo

Author: What does this app do?

Reviewer: This is an opensourced istio example that comes with 4 microservices, different versions to test out mesh capabilities. 

It will display information about a book. 

More details can be found here: https://istio.io/latest/docs/examples/bookinfo/

Reviewer: yeah that's a good idea, we just need to get off of evaluation mode with a license to have enough sidecars for the kuma app. I can do a quick test on openshift to make sure there are no other issues.

Author: Thanks!

Reviewer: I am concerned the postgres pod of the kuma demo is going to fail on openshift without some modification. Is there another usable demo?

1. Allow all the service accounts to use the kong-mesh-sidecar scc:
    ```bash
    oc adm policy add-scc-to-user kong-mesh-sidecar system:serviceaccount:bookinfo:bookinfo-details
    oc adm policy add-scc-to-user kong-mesh-sidecar system:serviceaccount:bookinfo:bookinfo-productpage
    oc adm policy add-scc-to-user kong-mesh-sidecar system:serviceaccount:bookinfo:bookinfo-ratings
    oc adm policy add-scc-to-user kong-mesh-sidecar system:serviceaccount:bookinfo:bookinfo-reviews
    ```

1. Next, deploy bookinfo. This bookinfo app has been paired down to work with Kong Mesh in evaluation mode. We only have 5 sidecars we can deploy on evaluation mode is all.
    ```bash
    kubectl create namespace bookinfo
    kubectl label namespace bookinfo kuma.io/sidecar-injection=enabled
    kubectl apply -f bookinfo/bookinfo.yaml -n bookinfo
    ```

1. Last we want to create an ingress resource to expose bookinfo to the outside world.
    ```bash
    kubectl apply -f bookinfo/ingress-productpage.yaml -n bookinfo
    ```

> 1. Now! Validate it's all up and running. Navigate to your browswer to `http://$PROXY_IP/productpage` and you should be able to see productpage with a majority of the bells and whistles.

Author: Is the purpose of using bookinfo here to see this page? Or are there other things customers can do with bookinfo now that it's hooked up to Mesh? I just want to make the connection between what customers would do here with bookinfo and what they would do in their own production environments.

Reviewer: They can test out more Kong Mesh features. I stopped here for the validation because this stopping point confirmed that all pods and sidecars are healthy.

> ## Clean Up 

Author: I like the addition of the cleanup section :)

---
title: Get started with Red Hat OpenShift and Kong Mesh
content_type: tutorial
description: This guide explains how to get started on Kong Mesh with Red Hat OpenShift.
---

> In this guide, you will learn how to get {{site.mesh_product_name}} up and running quickly in standalone mode on [Red Hat OpenShift](https://www.redhat.com/technologies/cloud-computing/openshift). This tutorial assumes some base-level OpenShift knowledge.

Author: ```suggestion
In this guide, you will learn how to get {{site.mesh_product_name}} up and running quickly in [standalone mode](/mesh/{{page.kong_version}}/production/deployment/stand-alone/) on [Red Hat OpenShift](https://www.redhat.com/technologies/cloud-computing/openshift). This tutorial assumes some base-level OpenShift knowledge.
```

This tutorial doesn't require a license because {{site.mesh_product_name}} can start in evaluation mode, which allows you to have up to five data planes or sidecars. This provides just enough data planes to get comfortable with the product and test it out.

This quickstart tutorial covers:

* How to use the Red Hat Certified {{site.mesh_product_name}} Images
* How to implement the required OpenShift security context constraints (SCCs) for the `kong-mesh` sidecar

> * How to deploy a sample application, bookinfo, on the mesh and validate that the application is working

Author: ```suggestion
* How to deploy a sample application, `kuma-demo`, on the mesh and validate that the application is working
```

* How to use the sample application to test the features of {{site.mesh_product_name}}

## Prerequisites
* [ROSA CLI installed](https://docs.openshift.com/rosa/rosa_install_access_delete_clusters/rosa_getting_started_iam/rosa-installing-rosa.html) or another OpenShift 4.12 cluster with the ability to create LoadBalancer type Kubernetes Services

> * [An AWS account with ROSA permissions enabled](https://docs.openshift.com/rosa/rosa_install_access_delete_clusters/rosa_getting_started_iam/rosa-config-aws-account.html)

Reviewer: "A AWS account" ?

* [kubectl CLI installed](https://kubernetes.io/docs/tasks/tools/)
* [OpenShift Container Platform CLI installed](https://docs.openshift.com/container-platform/latest/cli_reference/openshift_cli/getting-started-cli.html)
* [Helm 3 installed](https://helm.sh/docs/intro/install/)
* Clone the {{site.mesh_product_name}} OpenShift quickstart repository:
    ```bash

>     git clone https://github.com/Kong/kong-mesh-quickstart-openshift.git

Reviewer: do you need to tell the reader to navigate into the quickstart repo?

    ```

> * Clone the `kuma-demo` app:

Reviewer: Remove Clone `kuma-demo` bullet, because `kuma-demo` that will work on openshift is in the `kong-mesh-quickstart` repo

1. Create system variables for the cluster name and region:
    ```bash
    CLUSTER_NAME=kong-mesh-2
    REGION=us-west-2
    ```

1. Create a small ROSA cluster:
    ```bash
    rosa create cluster --cluster-name=$CLUSTER_NAME --region=$REGION --multi-az=false --version 4.12.13

>     ```

Reviewer: I did put extremely abbreviated instructions bc openshift customers will each manage rosa differently. 

With that said, did you first configure your AWS account? 
https://docs.openshift.com/rosa/rosa_install_access_delete_clusters/rosa_getting_started_iam/rosa-config-aws-account.html

> 1. Create system variables for the cluster name and region:

Author: Do everything from the cloned repo

    ```

1. When the cluster install is complete, create a cluster-admin user:
    ```bash
    rosa create admin --cluster $CLUSTER_NAME
    ```
    Creating an admin user gives you quick access to the cluster for this quickstart. We recommend using a formal identity provider that you grant admin privileges to in a production environment.

1. Validate you can log in to the cluster via the credentials provided by the ROSA CLI `stout`. Once you successfully login, you can proceed to the next section.

## Install {{site.mesh_product_name}} in standalone mode

In this section, you'll install {{site.mesh_product_name}} in standalone mode. Standalone mode is used in this quickstart for simplicity. 

1. Create the `kong-mesh-system` namespace:
    ```bash
    kubectl create namespace kong-mesh-system
    ```

1. Create the image pull secret:
    ```bash
    kubectl create secret docker-registry rh-registry-secret -n kong-mesh-system \
        --docker-server=registry.connect.redhat.com \
        --docker-username=<username> \
        --docker-password=<password> \
        --docker-email=<email>
    ```

>     Doing this allows you to authenticate so you can pull down the images from Red Hat's repository.

Reviewer: consider rephrase: "Doing this authenticates you to Red Hat's image registry, which allows you to pull the certified images"

    ```

    You should be able to reach the {{site.mesh_product_name}} UI at `http://localhost:5681/gui`.

1. Finally, you must do some prep work for the sidecar itself so sidecars will start up successfully. Apply the `kong-mesh-sidecar` SCCs and corresponding container patches:
    ```bash
    kubectl create -f kong-mesh/kong-mesh-sidecar-scc.yaml
    kubectl apply -f kong-mesh/container-patch.yaml 
    ```

    In OpenShift, by default, all pods and containers are given the restricted SCC. This is insufficient for the {{site.mesh_product_name}} sidecars (containers). The sidecars need a slightly escalated permissions, these permissions are defined in the `kong-mesh-sidecar-scc` manifest. The `container-patch` file defines how to patch the sidecars with the SCCs.

## Deploy the demo application

In this step, you'll deploy the [`kuma-demo` app](/mesh/{{page.kong_version}}/quickstart/kubernetes) to {{site.mesh_product_name}}. This allows you to quickly populate your mesh with services so you can test the capabilities of {{site.mesh_product_name}}.

The `kuma-demo` app consists of two services:

* `demo-app`: A web application that lets you increment a numeric counter

> * `redis`: A data store for the counter
> 
> 

Author: ```suggestion
* `redis`: A data store for the counter
```

1. Escalate the SCC of PostgreSQL SA to `privileged`:

  ```bash
  oc adm policy add-scc-to-user privileged -z kuma-demo-postgres -n kuma-demo
  ```

>   Escalating the SCC allows you to ?. This is only recommended for this tutorial, do not escalate the SSC of PostgreSQL SA in a production environment.

Reviewer: "Allows access to all privileged and host features"

  ```bash
  kubectl apply -f kuma-demo-aio-ocp.yaml
  ```

1. You should see four application pods, each with two containers, in the output:

  ```bash
  kuma-demo  $ oc get pods
  NAME                                    READY   STATUS    RESTARTS   AGE
  kuma-demo-app-5f5f455864-7bn5k          1/2     Running   0          9s
  kuma-demo-backend-v0-68bc879754-hck4t   1/2     Running   0          10s
  postgres-master-6576d699c8-f5wjn        2/2     Running   0          9s
  redis-master-946d4d567-8fvn5            2/2     Running   0          10s
  ```

1. Use `port-forward` to view the application:

  ```bash
  kubectl port-forward service/frontend -n kuma-demo 8080
  ```

  You can see the applications in the {{site.mesh_product_name}} UI at `http://localhost:8080/gui`.

## Test {{site.mesh_product_name}} capabilities

Now that you've deployed {{site.mesh_product_name}} along with the demo application on your ROSA cluster, you can test {{site.mesh_product_name}}. The following list captures some testing ideas for you to try:

* [Enable Mutual TLS and Traffic Permissions](/mesh/{{page.kong_version}}/quickstart/kubernetes/#enable-mutual-tls-and-traffic-permissions)

> * ?

Reviewer: What you've linked is good

    helm repo add kong-mesh https://kong.github.io/kong-mesh-charts
    ```

1. Install {{site.mesh_product_name}}:
    ```bash
    helm upgrade -i kong-mesh kong-mesh/kong-mesh -f kong-mesh/values.yaml -n kong-mesh-system
    ```

1. In a new terminal window, port-forward to reach the {{site.mesh_product_name}} GUI:
    ```bash
    kubectl port-forward svc/kong-mesh-control-plane -n kong-mesh-system 5681:5681
    ```

    You should be able to reach the {{site.mesh_product_name}} UI at `http://localhost:5681/gui`.

1. Finally, you must do some prep work for the sidecar itself so sidecars will startup successfully. Apply the `kong-mesh-sidecar` SCCs and corresponding container patches:
    ```bash
    kubectl create -f kong-mesh/kong-mesh-sidecar-scc.yaml
    kubectl apply -f kong-mesh/container-patch.yaml 
    ```

    In OpenShift, by default, all pods and containers are given the restricted SCC. This is insufficient for the {{site.mesh_product_name}} sidecars (containers). The sidecars need a slightly escalated permissions, these permissions are defined in the `kong-mesh-sidecar-scc` manifest. The `container-patch` file defines how to patch the sidecars with the SCCs.

> ## Deploy KIC on {{site.mesh_product_name}}

Reviewer: Remove the KIC section to allow the kuma-demo to work

    oc adm policy add-scc-to-user kong-mesh-sidecar system:serviceaccount:bookinfo:bookinfo-ratings
    oc adm policy add-scc-to-user kong-mesh-sidecar system:serviceaccount:bookinfo:bookinfo-reviews
    ```

1. Next, deploy bookinfo. This bookinfo app has been paired down to work with {{site.mesh_product_name}} in evaluation mode. We only have 5 sidecars we can deploy on evaluation mode is all.
    ```bash
    kubectl create namespace bookinfo
    kubectl label namespace bookinfo kuma.io/sidecar-injection=enabled
    kubectl apply -f bookinfo/bookinfo.yaml -n bookinfo
    ```

1. Last we want to create an ingress resource to expose bookinfo to the outside world.
    ```bash
    kubectl apply -f bookinfo/ingress-productpage.yaml -n bookinfo
    ```

1. Now! Validate it's all up and running. Navigate to your browswer to `http://$PROXY_IP/productpage` and you should be able to see productpage with a majority of the bells and whistles.

## List of things to try now that the demo is installed, just as a jumping off point and then the customers can further their own research/testing

## Clean Up 

> 1. Tear Down bookinfo:

Reviewer: Remove tear down bookinfo section, replace with `Tear Down Kuma-demo`

https://github.com/Kong/kong-mesh-quickstart-openshift/blob/main/README.md?plain=1#L130C5-L130C5

    ```bash
    kubectl delete deploy,svc,ingress --all -n bookinfo
    ```

> 1. Tear Down KIC:

Reviewer: Remove Tear Down KIC

---
title: Get started with Red Hat OpenShift and Kong Mesh
content_type: tutorial
description: This guide explains how to get started on Kong Mesh with Red Hat OpenShift.
---

In this guide, you will learn how to get {{site.mesh_product_name}} up and running quickly in [standalone mode](/mesh/{{page.kong_version}}/production/deployment/stand-alone/) on [Red Hat OpenShift](https://www.redhat.com/technologies/cloud-computing/openshift). This tutorial assumes some base-level OpenShift knowledge.

This tutorial doesn't require a license because {{site.mesh_product_name}} can start in evaluation mode, which allows you to have up to five data planes or sidecars. This provides just enough data planes to get comfortable with the product and test it out.

This quickstart tutorial covers:

* How to use the Red Hat Certified {{site.mesh_product_name}} images
* How to implement the required OpenShift security context constraints (SCCs) for the `kong-mesh` sidecar
* How to deploy a sample application, `kuma-demo`, on the mesh and validate that the application is working
* How to use the sample application to test the features of {{site.mesh_product_name}}

## Prerequisites

> * [ROSA CLI installed](https://docs.openshift.com/rosa/rosa_install_access_delete_clusters/rosa_getting_started_iam/rosa-installing-rosa.html) or another OpenShift 4.12 cluster with the ability to create LoadBalancer type Kubernetes Services

Author: ```suggestion
* [Red Hat OpenShift Service on AWS (ROSA) CLI installed](https://docs.openshift.com/rosa/rosa_install_access_delete_clusters/rosa_getting_started_iam/rosa-installing-rosa.html) or another OpenShift 4.12 cluster with the ability to create LoadBalancer type Kubernetes Services
```

* [An AWS account with ROSA permissions enabled](https://docs.openshift.com/rosa/rosa_install_access_delete_clusters/rosa_getting_started_iam/rosa-config-aws-account.html)
* [kubectl CLI installed](https://kubernetes.io/docs/tasks/tools/)
* [OpenShift Container Platform CLI installed](https://docs.openshift.com/container-platform/latest/cli_reference/openshift_cli/getting-started-cli.html)
* [Helm 3 installed](https://helm.sh/docs/intro/install/)
* Clone the {{site.mesh_product_name}} OpenShift quickstart repository:
    ```bash
    git clone https://github.com/Kong/kong-mesh-quickstart-openshift.git
    ```

## Install ROSA

> In this section, you will install a ROSA cluster called `kong-mesh-2` in the `us-west-2` region. Then, you'll create an admin user to get quick access to the cluster. 

Reviewer 2: Do we need to expand ROSA for this audience or not?

1. Navigate to the `kong-mesh-quickstart-openshift` repository that you just cloned in the prerequisites. All commands in this guide should be run from that repository.

1. Create system variables for the cluster name and region:
    ```bash
    CLUSTER_NAME=kong-mesh-demo
    REGION=us-west-2
    ```

1. Create a small ROSA cluster:
    ```bash
    rosa create cluster --cluster-name=$CLUSTER_NAME --region=$REGION --multi-az=false --version 4.12.13
    ```

1. When the cluster install is complete, create a cluster-admin user:
    ```bash
    rosa create admin --cluster $CLUSTER_NAME
    ```
    Creating an admin user gives you quick access to the cluster for this quickstart. We recommend using a formal identity provider that you grant admin privileges to in a production environment.

> 1. Validate you can log in to the cluster via the credentials provided by the ROSA CLI `stout`. Once you successfully login, you can proceed to the next section.

Reviewer 2: ```suggestion
1. Validate you can log in to the cluster via the credentials provided by the ROSA CLI `stout`. Once you successfully login, you can proceed to the next section.
```
```suggestion
1. Validate you can log in to the cluster using the credentials provided by the ROSA CLI `stout`. Once you successfully login, you can proceed to the next section.
```
